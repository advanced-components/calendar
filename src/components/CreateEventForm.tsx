import React, { useState } from "react";
import styled from "styled-components";

import { Button } from "./Button";
import { Form, FormField, Input } from "./Form";
import { getToday, getFullDate, getTime } from "../utils";

export const CreateEventForm = (props: any) => {
  const date = getToday();
  const getEvents = () => JSON.parse(localStorage.getItem("events") || "[]");
  const [events, setEvents] = useState(getEvents());
  const [eventValue, setEventValue] = useState({
    title: "",
    startTime: getTime(date, "startTime"),
    startDate: getFullDate(date),
    endTime: getTime(date, "endTime"),
    endDate: getFullDate(date),
    height: 39
  });
  const handleEventChange = (e: { target: HTMLInputElement }) => {
    const name = e.target.name;
    const value = e.target.value;
    setEventValue({ ...eventValue, [name]: value });
  };
  const addHourTime = Number(eventValue.endTime.slice(3, 5)) > 30 ? 1 : 0;
  const height =
    (Number(eventValue.endTime.slice(0, -3)) +
      addHourTime -
      Number(eventValue.startTime.slice(0, -3))) *
    39;
  const handleAddEvent = () => {
    const event = {
      title: eventValue.title == "" ? "Нет заголовка" : eventValue.title,
      startTime: eventValue.startTime,
      endTime: eventValue.endTime,
      startDate: eventValue.startDate,
      endDate: eventValue.endDate,
      height: height
    };
    setEvents([...events, event]);
    localStorage.setItem("events", JSON.stringify([...events, event]));
    props.onCloseModal();
  };

  return (
    <StyledCreateEventForm>
      <Form>
        <FormField label="Title">
          <Input
            type="text"
            placeholder="Event Title"
            name="title"
            data-testid="eventInput"
            onChange={e => handleEventChange(e)}
          />
        </FormField>
        <FormField label="Starts">
          <EventDate
            name="startDate"
            type="date"
            data-testid="eventInput"
            defaultValue={getFullDate(date)}
            onChange={e => handleEventChange(e)}
          ></EventDate>
          <EventTime
            name="startTime"
            type="time"
            data-testid="eventInput"
            defaultValue={getTime(date, "startTime")}
            onChange={e => handleEventChange(e)}
          ></EventTime>
        </FormField>
        <FormField label="Ends">
          <EventDate
            name="endDate"
            type="date"
            data-testid="eventInput"
            defaultValue={getFullDate(date)}
            onChange={e => handleEventChange(e)}
          ></EventDate>
          <EventTime
            name="endTime"
            type="time"
            data-testid="eventInput"
            defaultValue={getTime(date, "endTime")}
            onChange={e => handleEventChange(e)}
          ></EventTime>
        </FormField>
        <FormField>
          <div />
          <Button data-testid="save" onClick={() => handleAddEvent()}>
            Save
          </Button>
        </FormField>
      </Form>
    </StyledCreateEventForm>
  );
};

const StyledCreateEventForm = styled.div`
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 32px;
`;

const EventDate = styled(Input)`
  flex-grow: 1;
  margin-right: 8px;
`;

const EventTime = styled(Input)`
  flex-grow: 1;
`;
