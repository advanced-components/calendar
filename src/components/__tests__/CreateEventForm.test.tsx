import React from "react";
import { render, fireEvent, cleanup } from "react-testing-library";
import "jest-styled-components";
import "jest-dom/extend-expect";
import { CreateEventForm } from "../CreateEventForm";

afterEach(cleanup);

describe("CreateEventForm component", () => {
  it("render correctly", () => {
    const { container } = render(<CreateEventForm></CreateEventForm>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("change eventForm", () => {
    const onCloseModal = jest.fn();
    const { getByTestId } = render(
      <CreateEventForm onCloseModal={onCloseModal} />
    );
    const contentInput = getByTestId("eventInput");
    const submitButton = getByTestId("save");
    //@ts-ignore
    contentInput.value = "new content";
    fireEvent.change(contentInput);
    fireEvent.click(submitButton);
    //@ts-ignore
    expect(contentInput.value).toEqual("new content");
  });

  //@TODO test date inputs
});
