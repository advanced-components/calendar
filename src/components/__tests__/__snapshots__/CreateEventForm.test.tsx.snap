// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateEventForm component render correctly 1`] = `
.c8 {
  display: inline-block;
  zoom: 1;
  line-height: normal;
  white-space: nowrap;
  vertical-align: middle;
  text-align: center;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline: none;
  font-size: 13px;
  padding: 0.5em 1em;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-align: center;
}

.c8:hover {
  background-image: linear-gradient( transparent, rgba(0,0,0,0.05) 40%, rgba(0,0,0,0.1) );
}

.c8.active {
  background-image: linear-gradient( transparent, rgba(0,0,0,0.05) 40%, rgba(0,0,0,0.1) );
}

.c1 {
  min-width: 320px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  color: #777;
  font-size: 0.8em;
  position: relative;
  width: 100%;
}

.c5 {
  width: 100%;
  height: 35px;
  background-color: #fff;
  padding: 0 4px;
  border-radius: 3px;
  margin-top: 8px;
  outline: none;
  border: 1px solid #ccc;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 16px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 {
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 32px;
}

.c6 {
  width: 100%;
  height: 35px;
  background-color: #fff;
  padding: 0 4px;
  border-radius: 3px;
  margin-top: 8px;
  outline: none;
  border: 1px solid #ccc;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin-right: 8px;
}

.c7 {
  width: 100%;
  height: 35px;
  background-color: #fff;
  padding: 0 4px;
  border-radius: 3px;
  margin-top: 8px;
  outline: none;
  border: 1px solid #ccc;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

<div
  class="c0"
>
  <form
    class="c1"
  >
    <div
      class="c2"
    >
      <label
        class="c3"
      >
        Title
      </label>
      <div
        class="c4"
      >
        <input
          class="c5"
          data-testid="eventInput"
          name="title"
          placeholder="Event Title"
          type="text"
        />
      </div>
    </div>
    <div
      class="c2"
    >
      <label
        class="c3"
      >
        Starts
      </label>
      <div
        class="c4"
      >
        <input
          class="c6"
          data-testid="eventInput"
          name="startDate"
          type="date"
          value="2019-10-04"
        />
        <input
          class="c7"
          data-testid="eventInput"
          name="startTime"
          type="time"
          value="12:00"
        />
      </div>
    </div>
    <div
      class="c2"
    >
      <label
        class="c3"
      >
        Ends
      </label>
      <div
        class="c4"
      >
        <input
          class="c6"
          data-testid="eventInput"
          name="endDate"
          type="date"
          value="2019-10-04"
        />
        <input
          class="c7"
          data-testid="eventInput"
          name="endTime"
          type="time"
          value="13:00"
        />
      </div>
    </div>
    <div
      class="c2"
    >
      <div
        class="c4"
      >
        <div />
        <button
          class="c8"
          data-testid="save"
        >
          Save
        </button>
      </div>
    </div>
  </form>
</div>
`;
